dallinger/docker/tools.py

```diff
@@ -0,0 +1,188 @@
+import docker
+import os
+import shutil
+import time
+
+from jinja2 import Template
+from pathlib import Path
+from subprocess import check_output
+from subprocess import CalledProcessError
+
+from dallinger.utils import abspath_from_egg
+
+
+docker_compose_template = Template(
+    abspath_from_egg("dallinger", "dallinger/docker/docker-compose.yml.j2").read_text()
+)
+
+
+class DockerComposeWrapper(object):
+    """Wrapper around a docker compose local daemon, modeled after HerokuLocalWrapper.
+
+    Provides for verified startup and shutdown, and allows observers to register
+    to recieve subprocess output via 'monitor()'.
+
+    Implements a context manager pattern:
+
+        with DockerComposeWrapper(config, output) as docker:
+            docker.monitor(my_callback)
+
+    Arg 'output' should implement log(), error() and blather() methods taking
+    strings as arguments.
+    """
+
+    shell_command = "docker-compose"
+    MONITOR_STOP = object()
+
+    def __init__(
+        self,
+        config,
+        output,
+        original_dir,
+        tmp_dir,
+        verbose=True,
+        env=None,
+        needs_chrome=False,
+    ):
+        self.config = config
+        self.out = output
+        self.verbose = verbose
+        self.env = env if env is not None else os.environ.copy()
+        self.tmp_dir = tmp_dir
+        self.original_dir = original_dir
+        self.experiment_name = Path(self.original_dir).name
+        self._record = []
+        self.needs_chrome = needs_chrome
+        self.client = docker.APIClient(base_url="unix://var/run/docker.sock")
+
+    def copy_docker_compse_files(self):
+        for filename in ["Dockerfile.web", "Dockerfile.worker"]:
+            path = abspath_from_egg("dallinger", f"dallinger/docker/{filename}")
+            shutil.copy2(path, self.tmp_dir)
+        volumes = [f"{self.original_dir}:{self.original_dir}"]
+        with open(os.path.join(self.tmp_dir, "docker-compose.yml"), "w") as fh:
+            fh.write(
+                docker_compose_template.render(
+                    volumes=volumes,
+                    experiment_name=self.experiment_name,
+                    needs_chrome=self.needs_chrome,
+                )
+            )
+        with open(os.path.join(self.tmp_dir, ".env"), "w") as fh:
+            fh.write(f"COMPOSE_PROJECT_NAME=${self.experiment_name}\n")
+            fh.write(f"UID={os.getuid()}\n")
+            fh.write(f"GID={os.getgid()}\n")
+
+    def __enter__(self):
+        return self.start()
+
+    def start(self):
+        self.copy_docker_compse_files()
+        env = {"DOCKER_BUILDKIT": "1"}
+        build_arg = ""
+        if self.needs_chrome:
```

Comment:
This should probably look at the config variables `webdriver` and `recruiter` (there's a `"bots"` recruiter which is what `--bot` enables) to determine if it needs to install chrome or firefox, but I don't think that's critical at this stage. It gets more complicates with the `MultiRecruiter`, but I don't think we need to spend time on that right now.

Reply:
I think I disagree. I like the fact that the higher-level component (the `DockerDebugDeployment`) knows _why_ Chrome is needed and just tells the lower-level component (this one) to make it happen. The abstraction boundary leaks a little bit since the docker image is called "dallinger-bot", but nothing's perfect. :)

Reply:
You're right, but that I think that just means that at the `DockerDebugDeployment` we should be checking the `webdriver_type` and `recruiter` instead of relying on the `--bot` flag. For now it's probably fine as it is.

Reply:
Ah... yep; agreed.

Reply:
Got it, but leaving as is for now.